Deploy docs overview improvements (#15494)
PrefectHQ/prefect: docs/3.0/deploy/index.mdx

- **Schedule** or **Trigger**: optional schedule or triggering rules for this deployment
- **Tags**: optional text labels for organizing deployments

Some of the most common reasons to use an orchestration tool such as Prefect are for
[scheduling](/3.0/automate/add-schedules/) and [event-based triggering](/3.0/automate/events/automations-triggers/).
In addition to manually triggering and managing flow runs, deploying a flow exposes an API and
UI that allow you to:

- trigger new runs, [cancel active runs](/3.0/develop/write-flows/#cancel-a-flow-run),
pause scheduled runs, customize parameters, and more
- remotely configure [schedules](/3.0/automate/add-schedules) and [automation rules](/3.0/automate/events/automations-triggers) for deployments
- dynamically provision infrastructure with [work pools](/3.0/deploy/infrastructure-concepts/work-pools)

## Create a deployment 

### Create a deployment in Python with a local `serve` method

> This is a great way to get started with deployments.

Reviewer: ```suggestion
This is the simplest way to get started with deployments:
```

```python
from prefect import flow

@flow
def my_flow():
    print("Hello, Prefect!")

if __name__ == "__main__":
    my_flow.serve(name="my-first-deployment")
```

> This methods creates a deployment object that can be used to trigger flow runs.
> It starts a server and registers the flow with the Prefect API.
> The process runs locally and waits to pick up scheduled flow runs from the server.

Reviewer: I think we should update the language here. in particular
```suggestion
This method creates a deployment from your flow and immediately begins listening for scheduled runs to execute.
```
using "server" here like this is not quite accurate

> If you want to run your flow on remote infrastructure, need more control over the execution environment, or want to create a template for other deployments, use a deployment with a work pool.

Reviewer: perhaps these criteria could be bullets?

Author: Updated and pushed - also made a few other changes.

of the infrastructure on which flows run.
Options for [serverless work pools](/3.0/deploy/infrastructure-examples/serverless/) allow you to
scale to zero when workflows aren't running.
Prefect even provides you with the ability to
[provision cloud infrastructure via a single CLI command](/3.0/deploy/infrastructure-examples/serverless/#automatically-create-a-new-push-work-pool-and-provision-infrastructure),
if you use a Prefect Cloud push work pool option.

With work pools:

- You can configure and monitor infrastructure configuration within the Prefect UI.
- Infrastructure is ephemeral and dynamically provisioned.
- Prefect is more infrastructure-aware and collects more event data from your infrastructure by default.
- Highly decoupled setups are possible.

<Note>
**You don't have to commit to one approach**

You can mix and match approaches based on the needs of each flow. You can also change the
deployment approach for a particular flow as its needs evolve.
For example, you might use workers for your expensive machine learning pipelines,
but use the serve mechanics for smaller, more frequent file-processing pipelines.

> </Note>

Reviewer: nice note!

- **Schedule** or **Trigger**: optional schedule or triggering rules for this deployment
- **Tags**: optional text labels for organizing deployments

Some of the most common reasons to use an orchestration tool such as Prefect are for
[scheduling](/3.0/automate/add-schedules/) and [event-based triggering](/3.0/automate/events/automations-triggers/).
In addition to manually triggering and managing flow runs, deploying a flow exposes an API and
UI that allow you to:

- trigger new runs, [cancel active runs](/3.0/develop/write-flows/#cancel-a-flow-run),
pause scheduled runs, customize parameters, and more
- remotely configure [schedules](/3.0/automate/add-schedules) and [automation rules](/3.0/automate/events/automations-triggers) for deployments
- dynamically provision infrastructure with [work pools](/3.0/deploy/infrastructure-concepts/work-pools)

## Create a deployment 

> ### Create a deployment in Python with a local `serve` method

Reviewer 2: Suggestion for something a little pithier
```suggestion
### Create a deployment with `serve`
```

This is the simplest way to get started with deployments:

```python
from prefect import flow

@flow
def my_flow():
    print("Hello, Prefect!")

if __name__ == "__main__":
    my_flow.serve(name="my-first-deployment", cron="* * * * *")
```

The `serve` method creates a deployment from your flow and immediately begins listening for scheduled runs to execute. 

> It also create a schedule so that the flow will run every minute of every day.

Reviewer 2: ```suggestion
This example also create a schedule so that the flow will run every minute of every day.
```

> For slightly more configuration you can create a deployment that uses a work pool.
> Reasons to create a work-pool based deployment include:
> 
> - Wanting to run your flow dynamically on remote infrastructure
> - Needing more control over the execution environment
> - Creating a template for other deployments

Reviewer 2: ```suggestion
For more configuration, you can create a deployment that uses a work pool.
Reasons to create a work-pool based deployment include:

- Wanting to run your flow on dynamically provisioned infrastructure
- Needing more control over the execution environment on a per-flow run basis
- Creating an infrastructure template to use across deployments
```

Work pools are popular with data platform teams because they allow you to manage infrastructure configuration across an organization. 

### Create a work pool-based deployment

> Prefect offers two primary options for creating deployments that use a work pool: a Python method and a YAML file.

Reviewer 2: ```suggestion
Prefect offers two options for creating work pool-based deployments: a Python script or a YAML file.
```

> ### Create a deployment with a Python `flow.deploy` method

Reviewer 2: I suggest this name to mirror the serve section above, but it could be `flow.deploy` instead
```suggestion
### Create a deployment with `deploy`
```

> To define a deployment through Python code, use the `flow.deploy` method.

Reviewer 2: ```suggestion
To define a deployment with a Python script, use the `flow.deploy` method.
```

```bash
prefect deploy
```

The result is a `prefect.yaml` file for deployment creation. 
The file contains `build`, `push`, and `pull` steps for building a Docker image, pushing code to a Docker registry, and pulling code at runtime.
Learn more about creating deployments with a YAML file in [Define deployments with YAML](/3.0/deploy/infrastructure-concepts/prefect-yaml).

Prefect also provides [CI/CD options](/3.0/deploy/infrastructure-concepts/deploy-ci-cd) for automatically creating YAML-based deployments. 

### Work pools

[Work pools](/3.0/deploy/infrastructure-concepts/work-pools/) allow you to switch between different types of infrastructure and to create a template for deployments. 
Data platform teams find work pools especially useful for managing infrastructure configuration across teams of data professionals.

Common work pool types include [Docker](/3.0/deploy/infrastructure-examples/docker), [Kubernetes](/3.0/deploy/infrastructure-examples/kubernetes), and serverless options such as [AWS ECS](/integrations/prefect-aws/ecs_guide#ecs-worker-guide), [Azure ACI](/integrations/prefect-azure/aci_worker), and [GCP Cloud Run](/integrations/prefect-gcp/gcp-worker-guide). 

> ### Work pool-based deployment requirements
> 
> Deployments created through the Python SDK that use a work pool require a `name` argument and a `work_pool` argument. 
> They also require reference to your flow code storage location.
> 
> Your flow code location can be specified in a few ways:
> 
> 1. Bake it into your Docker image (for work-pools that use Docker images). 
>     As shown in the example above,Prefect facilitates this as the default method for deployments created with the Python SDK. 
>     This method requires that you specify the `image` argument in the `deploy` method.
> 2. Chain the `from_source` method to a flow object or flow function and specify one of the following:
>     1. the git-based cloud provider location (for example, GitHub) 
>     2. the cloud provider storage location (for example, AWS S3)
>     3. the local path (an option for Process work pools)
> 
> See the [Retrieve code from storage docs](/3.0/deploy/infrastructure-concepts/store-flow-code) for more information about flow code storage.

Reviewer 2: ```suggestion
### Work pool-based deployment requirements

Deployments created through the Python SDK that use a work pool require a reference to your flow code storage location. This information is used to download and run your code when your deployment is run.

Your flow code location can be specified in a few ways:

1. Bake it into your Docker image (for work-pools that use Docker images). 
    As shown in the example above,Prefect facilitates this as the default method for deployments created with the Python SDK. 
    This method requires that you specify the `image` argument in the `deploy` method.
2. Call `from_source` on a flow and specify one of the following:
    1. the git-based cloud provider location (for example, GitHub) 
    2. the cloud provider storage location (for example, AWS S3)
    3. the local path (an option for Process work pools)

See the [Retrieve code from storage docs](/3.0/deploy/infrastructure-concepts/store-flow-code) for more information about flow code storage.
```

These work pool types do not require a worker to run flows. 
However, they do require sharing a bit more information with Prefect, which can be a challenge depending upon the security posture of your organization.

## Static vs. dynamic infrastructure

You can deploy your flows on long-lived static infrastructure or on dynamic infrastructure that is able to scale horizontally.
The best choice depends on your use case.

### Static infrastructure

When you have several flows running regularly, [the `serve` method](/3.0/develop/write-flows/#serving-a-flow)
of the `Flow` object or [the `serve` utility](/3.0/develop/write-flows/#serving-multiple-flows-at-once)
is a great option for managing multiple flows simultaneously.

Once you have authored your flow and decided on its deployment settings, run this long-running
process in a location of your choosing.
The process stays in communication with the Prefect API, monitoring for work and submitting each run
within an individual subprocess.
Because runs are submitted to subprocesses, any external infrastructure configuration
must be set up beforehand and kept associated with this process.

> This approach has these benefits:

Reviewer 2: ```suggestion
Benefits to this approach include:
```

Deployment metadata references code that exists in potentially diverse locations within your environment.
This separation means that your flow code stays within your storage and execution
infrastructure.

> This is key to the Prefect hybrid model used with hybrid work pool types: there's a boundary between your proprietary assets,

Reviewer: i personally think this is more clear as it was
```suggestion
This is key to the Prefect hybrid model: there's a boundary between your proprietary assets,
```
